Add C# code to random_number_generation and fix shuffle bag example. (#8072)
godotengine/godot-docs: tutorials/math/random_number_generation.rst

 .. code-tab:: gdscript GDScript

    # Prints a random floating-point number from a normal distribution with a mean 0.0 and deviation 1.0.
    print(randfn(0.0, 1.0))

 .. code-tab:: csharp

>     // Prints a normally distributed floating-point number between 0.0 and 1.0.

Reviewer: ```suggestion
    // Prints a random floating-point number from a normal distribution with a mean 0.0 and deviation 1.0.
```

.. tabs::
 .. code-tab:: gdscript GDScript

    var _metals = {
        "copper": {"quantity": 50, "price": 50},
        "silver": {"quantity": 20, "price": 150},
        "gold": {"quantity": 3, "price": 500},
    }

    func _ready():
        for i in range(20):
            print(get_metal())

    func get_metal():
        var random_metal = _metals.values()[randi() % metals.size()]
        # Returns a random metal value dictionary every time the code runs.
        # The same metal may be selected multiple times in succession.
        return random_metal

 .. code-tab:: csharp

    private Godot.Collections.Dictionary<string, Godot.Collections.Dictionary<string, int>> _metals = new()
    {

>         {"copper", new Godot.Collections.Dictionary{{"quantity", 50}, {"price", 50}}},
>         {"silver", new Godot.Collections.Dictionary{{"quantity", 20}, {"price", 150}}},
>         {"gold", new Godot.Collections.Dictionary{{"quantity", 3}, {"price", 500}}}

Reviewer: ```suggestion
        {"copper", new Godot.Collections.Dictionary<string, int>{{"quantity", 50}, {"price", 50}}},
        {"silver", new Godot.Collections.Dictionary<string, int>{{"quantity", 20}, {"price", 150}}},
        {"gold", new Godot.Collections.Dictionary<string, int>{{"quantity", 3}, {"price", 500}}},
```

    };

    public override void _Ready()
    {
        for (int i = 0; i < 20; i++)
        {
            GD.Print(GetMetal());
        }
    }

>     public Godot.Collections.Dictionary GetMetal()

Reviewer: ```suggestion
    public Godot.Collections.Dictionary<string, int> GetMetal()
```

    {

>         var (_, randomMetal) = _metals.ElementAt(GD.Randi() % metals.Count);

Reviewer: ```suggestion
        var (_, randomMetal) = _metals.ElementAt((int)(GD.Randi() % _metals.Count));
```

    func get_metal():
        var random_metal = _metals.values()[randi() % metals.size()]
        # Returns a random metal value dictionary every time the code runs.
        # The same metal may be selected multiple times in succession.
        return random_metal

 .. code-tab:: csharp

    private Godot.Collections.Dictionary _metals = new Godot.Collections.Dictionary
    {
        {"copper", new Godot.Collections.Dictionary{{"quantity", 50}, {"price", 50}}},
        {"silver", new Godot.Collections.Dictionary{{"quantity", 20}, {"price", 150}}},
        {"gold", new Godot.Collections.Dictionary{{"quantity", 3}, {"price", 500}}}
    };

    public override void _Ready()
    {

>         GD.Randomize();

Reviewer 2: TIL 😲

        for (int i = 0; i < 20; i++)
        {
            GD.Print(GetMetal());
        }
    }

    public Godot.Collections.Dictionary GetMetal()
    {

>         var randomMetal = _metals.KeyValuePair(GD.Randi() % metals.Count);

Reviewer 2: Wouldn't work either, `Godot.Collections.Dictionary.Values` gives an `ICollection` (doesn't implement `this[int]`).

>         var randomMetal = _metals.KeyValuePair(GD.Randi() % metals.Count);

Reviewer: Oh, you are right. I missed that.

>         var randomMetal = _metals.KeyValuePair(GD.Randi() % metals.Count);

Author: Oh shoot, I meant [GetKeyValuePair().Value](https://github.com/godotengine/godot/blob/19890614c6a78ec36030ce65c7da05f07fcdb9ed/modules/mono/glue/GodotSharp/GodotSharp/Core/Dictionary.cs#L428)

(Should stop doing quick edits half awake)

>         var randomMetal = _metals.KeyValuePair(GD.Randi() % metals.Count);

Reviewer 2: `GetKeyValuePair()` is private. I don't think there's really a straightforward way to oneline this using `Godot.Collections.Dictionary`.

>         var randomMetal = _metals.KeyValuePair(GD.Randi() % metals.Count);

Reviewer: I guess the best we could do here is use Linq's [`ElementAt`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.elementat), if we want to keep it short.

```csharp
var (_, randomMetal) = _metals.ElementAt(GD.Randi() % metals.Count);
return randomMetal;
```

        # The same metal may be selected multiple times in succession.
        return random_metal

 .. code-tab:: csharp

>     private var metals = new Godot.Collections.Dictionary

Reviewer 2: `var` can't be used to define a field.

    {
        {"copper", new Godot.Collections.Dictionary{{"quantity", 50}, {"price", 50}}},
        {"silver", new Godot.Collections.Dictionary{{"quantity", 20}, {"price", 150}}},
        {"gold", new Godot.Collections.Dictionary{{"quantity", 3}, {"price", 500}}}
    };

    public override void _Ready()
    {
        GD.Randomize();
        for (int i = 0; i < 20; i++)
        {
            GD.Print(GetMetal());
        }
    }

    public Godot.Collections.Dictionary GetMetal()
    {

>         var randomMetal = metals.Values()[GD.Randi() % metals.Length];

Reviewer 2: This method doesn't return.
This line wouldn't compile for several reasons:
- `.Values` is a property, not a method
- I don't think the return of `.Values` can be `[]`-indexed
- Godot collections use `.Count` and not `.Length`

    func get_fruit():
        if _fruits.is_empty():
            # Fill the fruits array again and shuffle it.
            _fruits = _fruits_full.duplicate()
            _fruits.shuffle()

        # Get a random fruit, since we shuffled the array,
        # and remove it from the `_fruits` array.
        var random_fruit = _fruits.pop_front()
        # Returns "apple", "orange", "pear", or "banana" every time the code runs, removing it from the array.
        # When all fruit are removed, it refills the array.
        return random_fruit

 .. code-tab:: csharp

    private Godot.Collections.Array<string> _fruits = new() { "apple", "orange", "pear", "banana" };
    // A copy of the fruits array so we can restore the original value into `fruits`.

>     private Godot.Collections.Array _fruitsFull;

Reviewer: Since this is a copy of the `_fruits` array it should have the same type.

```suggestion
    private Godot.Collections.Array<string> _fruitsFull;
```

    public override void _Ready()
    {
        _fruitsFull = _fruits.Duplicate();
        _fruits.Shuffle();

        for (int i = 0; i < 100; i++)
        {
            GD.Print(GetFruit());
        }
    }

    public string GetFruit()
    {
        if(_fruits.Count == 0)
        {
            // Fill the fruits array again and shuffle it.
            _fruits = _fruitsFull.Duplicate();
            _fruits.Shuffle();
        }

        // Get a random fruit, since we shuffled the array,

>         var randomFruit = _fruits[0];

Reviewer: We should probably avoid using `var` in the documentation to make the type more clear.

```suggestion
        string randomFruit = _fruits[0];
```

        // and remove it from the `_fruits` array.
        _fruits.RemoveAt(0);
        // Returns "apple", "orange", "pear", or "banana" every time the code runs, removing it from the array.
        // When all fruit are removed, it refills the array.

>         return randomFruit.AsString();

Reviewer: `randomFruit` is already a string, not a Variant, because the array is typed.

```suggestion
        return randomFruit;
```

        # When all fruit are removed, it refills the array.
        return random_fruit

 .. code-tab:: csharp

    private var _fruits = new Godot.Collections.Array{ "apple", "orange", "pear", "banana" };
    // A copy of the fruits array so we can restore the original value into `fruits`.
    private Godot.Collections.Array _fruitsFull;

    public override void _Ready()
    {
        GD.Randomize();
        _fruitsFull = _fruits.Duplicate();
        _fruits.Shuffle();

        for (int i = 0; i < 100; i++)
        {
            GD.Print(GetFruit());
        }
    }

    public string GetFruit()
    {
        if(_fruits.Length == 0)
        {
            // Fill the fruits array again and shuffle it.
            _fruits = _fruitsFull.Duplicate();
            _fruits.Shuffle();
        }

        // Get a random fruit, since we shuffled the array,
        var randomFruit = _fruits[0];
        // and remove it from the `_fruits` array.
        _fruits.RemoveAt(0);
        // Returns "apple", "orange", "pear", or "banana" every time the code runs, removing it from the array.
        // When all fruit are removed, it refills the array.

>         return randomFruit;

Reviewer 2: If we're using `Godot.Collections.Array`, `randomFruit` is a `Variant` here. We either need to return `randomFruit.AsString()`, or use a generic collection `Godot.Collections.Array<string>` to begin with.

            # Fill the fruits array again and shuffle it.
            _fruits = _fruits_full.duplicate()
            _fruits.shuffle()

        # Get a random fruit, since we shuffled the array,
        # and remove it from the `_fruits` array.
        var random_fruit = _fruits.pop_front()
        # Returns "apple", "orange", "pear", or "banana" every time the code runs, removing it from the array.
        # When all fruit are removed, it refills the array.
        return random_fruit

 .. code-tab:: csharp

    private Godot.Collections.Array _fruits = new Godot.Collections.Array{ "apple", "orange", "pear", "banana" };
    // A copy of the fruits array so we can restore the original value into `fruits`.
    private Godot.Collections.Array _fruitsFull;

    public override void _Ready()
    {
        GD.Randomize();
        _fruitsFull = _fruits.Duplicate();
        _fruits.Shuffle();

        for (int i = 0; i < 100; i++)
        {
            GD.Print(GetFruit());
        }
    }

    public string GetFruit()
    {
        if(_fruits.Count == 0)
        {
            // Fill the fruits array again and shuffle it.
            _fruits = _fruitsFull.Duplicate();

>             _fruits.Shuffle();

Reviewer: If we use System collections, shuffling can be implemented as a custom method:

```csharp
private void Shuffle(Metal[] metals)
{
	for (int i = 0; i < metals.Length - 1; i++)
	{
		int j = GD.RandRange(i, metals.Length);
		if (j != i)
		{
			(metals[i], metals[j]) = (metals[j], metals[i]);
		}
	}
}
```

@paulloz Do you think this is good enough or would it be better to keep using Godot collections for the convenience of the `Shuffle` method?

Reviewer 2: It really depends on what we want to emphasize on this page. Unless we explicitly want to emphasize usage of `Godot.Collections`, I'd rather use system ones.  

On a somewhat unrelated note, we could implement a `GD.Shuffle()` (or somewhere else in the API) that uses the Godot random, and list it on the [API differences](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_differences.html) page at some point.

Reviewer: > I also think that any valid version of this is better than what we currently have

Agreed, so let's not block the PR on this. I'll resolve.

    public override void _Ready()
    {
        // Configure the FastNoiseLite instance.

>         _noise.NoiseType = NoiseTypeEnum.SimplexSmooth;

Reviewer: ```suggestion
        _noise.NoiseType = FastNoiseLitle.NoiseTypeEnum.SimplexSmooth;
```
